### ISSUE TITLE
Add option to specify caching directory

### ISSUE THREAD

**Issue by Moini** [phw/peek#138]
When exploring why there's so little space on my disk, (among other things) I found that peek has left a > 1GB file in its cache directory in my home directory (and has not removed it, maybe result of a crash or something).

Would it be possible to add an option for specifying the caching directory?  
If peek adds some more of these files, or puts a huge recording there while I'm using it, I'll run into problems...

Using peek 1.0.1 (just in case the option has been added since I last looked).

**Comment by phw** [comment]
I was hesitant to add such an option, as in the past people could just set `TMP_DIR`. But since Peek is now handling the cache directory itself I will probably add a hidden dconf option.

> I found that peek has left a > 1GB file in its cache directory in my home directory (and has not removed it, maybe result of a crash or something).

If there are leftover files in there just clean it up. After Peek finished recording this directory should be empty. If there is anything left it is likely due to some crash (or it is a bug).

**Comment by Moini** [comment]
Yes, that will be helfpul. Sometimes, due to other people also working on my computer, I only have very little disk memory left, and if I record anything where the cache file exceeds this remainder, I might run into difficulties.

**Comment by sQVe** [comment]
I for once, and many others (especially Arch users), need this option heavily as it is _quite common_ that `$HOME/.cache` is *tmpfs*. My setup for `$HOME/.cache` is: `/home/sqve/.cache tmpfs noatime,nodev,nosuid,size=500M`.

This makes peek crash, really quick, when 500MB is used. Adding an options or a environment variable to set the cache dir would solve this should a pretty quick and solid fix.

**Comment by phw** [comment]
Makes definitely sense in this setup. Still this is rather low on my priority list, the setup seems uncommon enough (it kinda defeats the purpose of the cache directory). But patches welcome :)

**Comment by sQVe** [comment]
@phw I'd love to create a patch but I'm very unfimiliar with Vala and don't have much spare time to play around with. I do not agree that the setup defeats the purpose of a cache directory though. The idea about having it in tmpfs is that it is way faster, saves writes on SSD and it is cleared on reboot etc. The only application that has had problems with this, and the limit of 500MB, is in fact *peek*, and I've been running this setup for 4+ years.

**Comment by phw** [comment]
> ...and it is cleared on reboot etc.

I don't mean to imply you shouldn't be using this setup, as long as you know what you do this is fine. With "it defeats the purpose of the cache directory" I just mean your usage nullifies some reasons why cache files are used in the first place.

Don't count Peek here. Peek just uses ~/.cache as temp file storage. Temp files are usually written for data you need to keep around for a short period of time, but the developer considers this data to be to large to hold in RAM. After being used it is thrown away again. Fast access is important here, hence it makes sense to have this in tmpfs as long as you know you have enough RAM to hold temporary data for the applications you use.

The typical cache file is different. Cache files are written for data that is a) expensive to recreate (e.g. obtained over slow network or created via processor intensive calculations) **and** b) the data can be again in the future (avoiding the recreation). If you put ~/.cache on a tmpfs you get the benefit of very fast cache access at the expense of frequent cache recreations. This makes the cache for some applications less useful or in some cases probably completely useless (e.g. if the application only uses the cache to speed up startup, and you start that application only once per session and then reboot).

Bottom line: You have to know what applications you use for the right choice to weight pros and cons, and that's the reason I consider this an unusual configuration for general purpose usage.

To offer a workaround: Why not symlink `~/.cache/peek` to some persistent location?

**Comment by sQVe** [comment]
@phw Yeah I fully agree that this setup is somewhat unusual. I'm fully aware on what this setup does and those are the reasons I run my setup like this. I prefer having my cache set for only one specific session, which I haven't had a problem with until now.

Using a symlink is a great suggestion. I will try that and see it that makes a difference. :beers:

**Comment by sQVe** [comment]
@phw That did not make a difference. I still get the following:

```
% peek                                                                                   
Using screen recorder backend ffmpeg
ffmpeg version 3.4 Copyright (c) 2000-2017 the FFmpeg developers
  built with gcc 7.2.0 (GCC)
  configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-avisynth --enable-avresample --enable-fontconfig --enable-gmp --enable-gnutls --enable-gpl --enable-ladspa --enable-libass --enable-libbluray --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libiec61883 --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxcb --enable-libxvid --enable-shared --enable-version3
  libavutil      55. 78.100 / 55. 78.100
  libavcodec     57.107.100 / 57.107.100
  libavformat    57. 83.100 / 57. 83.100
  libavdevice    57. 10.100 / 57. 10.100
  libavfilter     6.107.100 /  6.107.100
  libavresample   3.  7.  0 /  3.  7.  0
  libswscale      4.  8.100 /  4.  8.100
  libswresample   2.  9.100 /  2.  9.100
  libpostproc    54.  7.100 / 54.  7.100
[x11grab @ 0x555c96faab00] Stream #0: not enough frames to estimate rate; consider increasing probesize
Input #0, x11grab, from ':0+1286,829':
  Duration: N/A, start: 1509094297.797364, bitrate: N/A
    Stream #0:0: Video: rawvideo (BGR[0] / 0x524742), bgr0, 1268x605, 10 fps, 1000k tbr, 1000k tbn, 1000k tbc
Stream mapping:
  Stream #0:0 -> #0:0 (rawvideo (native) -> pam (native))
Press [q] to stop, [?] for help
Output #0, rawvideo, to '/home/sqve/.cache/peek/peekK26U8Y.pam':
  Metadata:
    encoder         : Lavf57.83.100
    Stream #0:0: Video: pam, rgb24, 1268x605, q=2-31, 200 kb/s, 10 fps, 10 tbn, 10 tbc
    Metadata:
      encoder         : Lavc57.107.100 pam
frame=   50 fps= 10 q=-0.0 Lsize=  112377kB time=00:00:05.00 bitrate=184118.7kbits/s speed=1.04x    
video:112377kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.000000%
ffmpeg version 3.4 Copyright (c) 2000-2017 the FFmpeg developers
  built with gcc 7.2.0 (GCC)
  configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-avisynth --enable-avresample --enable-fontconfig --enable-gmp --enable-gnutls --enable-gpl --enable-ladspa --enable-libass --enable-libbluray --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libiec61883 --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxcb --enable-libxvid --enable-shared --enable-version3
  libavutil      55. 78.100 / 55. 78.100
  libavcodec     57.107.100 / 57.107.100
  libavformat    57. 83.100 / 57. 83.100
  libavdevice    57. 10.100 / 57. 10.100
  libavfilter     6.107.100 /  6.107.100
  libavresample   3.  7.  0 /  3.  7.  0
  libswscale      4.  8.100 /  4.  8.100
  libswresample   2.  9.100 /  2.  9.100
  libpostproc    54.  7.100 / 54.  7.100
Input #0, image2, from '/home/sqve/.cache/peek/peekK26U8Y.pam':
  Duration: 00:00:00.04, start: 0.000000, bitrate: 23014840 kb/s
    Stream #0:0: Video: pam, rgb24, 1268x605, 25 tbr, 25 tbn, 25 tbc
Stream mapping:
  Stream #0:0 -> #0:0 (pam (native) -> png (native))
Press [q] to stop, [?] for help
Output #0, image2, to '/home/sqve/.cache/peek/peekJ7748Y.png':
  Metadata:
    encoder         : Lavf57.83.100
    Stream #0:0: Video: png, rgba, 16x16 [SAR 1:1 DAR 1:1], q=2-31, 200 kb/s, 10 fps, 10 tbn, 10 tbc
    Metadata:
      encoder         : Lavc57.107.100 png
frame=    0 fps=0.0 q=0.0 Lsize=N/A time=00:00:00.00 bitrate=N/A speed=   0x    
video:0kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
Output file is empty, nothing was encoded (check -ss / -t / -frames parameters if used)
ffmpeg version 3.4 Copyright (c) 2000-2017 the FFmpeg developers
  built with gcc 7.2.0 (GCC)
  configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-avisynth --enable-avresample --enable-fontconfig --enable-gmp --enable-gnutls --enable-gpl --enable-ladspa --enable-libass --enable-libbluray --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libiec61883 --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxcb --enable-libxvid --enable-shared --enable-version3
  libavutil      55. 78.100 / 55. 78.100
  libavcodec     57.107.100 / 57.107.100
  libavformat    57. 83.100 / 57. 83.100
  libavdevice    57. 10.100 / 57. 10.100
  libavfilter     6.107.100 /  6.107.100
  libavresample   3.  7.  0 /  3.  7.  0
  libswscale      4.  8.100 /  4.  8.100
  libswresample   2.  9.100 /  2.  9.100
  libpostproc    54.  7.100 / 54.  7.100
Input #0, image2, from '/home/sqve/.cache/peek/peekK26U8Y.pam':
  Duration: 00:00:00.04, start: 0.000000, bitrate: 23014840 kb/s
    Stream #0:0: Video: pam, rgb24, 1268x605, 25 tbr, 25 tbn, 25 tbc
/home/sqve/.cache/peek/peekJ7748Y.png: Invalid data found when processing input
Error: Child process exited with code 1
File save error: Error when getting information for file “/home/sqve/.cache/peek/peekPQVO8Y.gif”: No such file or directory
Temp file delete error: Error removing file /home/sqve/.cache/peek/peekPQVO8Y.gif: No such file or directory

```

**Comment by phw** [comment]
@sQVe This sounds more like https://github.com/phw/peek/issues/198 . The new GIF encoding with ffmpeg seems to make issues. Can you try with peek release version 1.1.0 instead?

**Comment by Moini** [comment]
Solved it with a symlink to from /home/username/.cache/peek to /tmp - worked for me.

**Comment by sQVe** [comment]
@phw Symlinking is working like a charm. I suggest closing this issue.

#198 is the exact issue I was having. When reverting back to AUR `peek` package instead of `peek-git` everything works like it should.

**Comment by phw** [comment]
The latest Peek from git should be rather low on resource usage, including the temporary space. Of course this can still grow to sizes too large for a RAM disk (depending on what you record), but in any way I am pretty confident this is not a too big deal anymore.

Anyway, I'll keep this issue open as you have convinced me that this needs to be configurable in some cases :)
